### PR TITLE
correct handling of travel modes

### PR DESCRIPTION
**Pull request by MoKob**
At the moment, travel modes are handled in a very unintuitive way.
The travel mode of a segment cannot be deduced from its own ID but only from the turn-id it has been reached from (and the phantom node itself).

This comes with additional implied behaviour for the turn-by-turn generation.

To allow fine grained weights (like in https://github.com/Project-OSRM/osrm-backend/pull/1970) along edges, we need the travel mode to not run into problems with this implied behaviour.

In addition, it allows us to do better instructions based on travel modes (e.g. in the tunnel, take the fork on the right) or similar stuff.

### CODE REVIEW

File: include/engine/routing_algorithms/routing_base.hpp [hunk 1]
```diff
@@ -675,11 +679,11 @@ template <class DataFacadeT, class Derived> class BasicRoutingInterface
         BOOST_ASSERT_MSG((SPECIAL_NODEID != middle && INVALID_EDGE_WEIGHT != distance),
                          "no path found");
 
-        if (distance != forward_heap.GetKey(middle) + reverse_heap.GetKey(middle))
```
Comment (TheMarex) [inline include/engine/routing_algorithms/routing_base.hpp]: This looks like an unrelated bug fix, can we break this out?

File: include/extractor/edge_based_node.hpp [hunk 2]
```diff
@@ -9,6 +9,7 @@
 #include "osrm/coordinate.hpp"
 
 #include <limits>
+#include <string>
```
Comment (TheMarex) [inline include/extractor/edge_based_node.hpp]: We don't need string here.

File: include/engine/guidance/segment_list.hpp [hunk 3]
```diff
@@ -235,6 +211,30 @@ void SegmentList<DataFacadeT>::Finalize(const bool extract_alternative,
...
+    }
+
+    // announce mode changes
+    for (std::size_t i = 0; i + 1 < segments.size(); ++i)
+    {
+        auto &segment = segments[i];
+        const auto next_mode = segments[i + 1].travel_mode;
+        if (segment.travel_mode != next_mode &&
+            segment.turn_instruction == extractor::TurnInstruction::NoTurn)
+        {
```
Comment (TheMarex) [inline include/engine/guidance/segment_list.hpp]: Couldn't we handle this in pre-processing in the turn generation?

Comment (MoKob) [inline include/engine/guidance/segment_list.hpp]: Not without adding additional information to the edges.

In turning from (a)->(b), we want to know both about the modes of (a) for all information stored along (a), as well as (b). The turn announces the mode change to the mode of (b) at (a). So you either have to store both the current and the next travel mode, or you have to traverse the path to find the travel mode.

File: include/engine/guidance/segment_list.hpp [hunk 4]
```diff
@@ -153,7 +154,7 @@ void SegmentList<DataFacadeT>::AddLeg(const std::vector<PathData> &leg_data,
                                       const bool is_via_leg,
                                       const DataFacade *facade)
 {
-    for (const PathData &path_data : leg_data)
+    for (const auto path_data : leg_data)
```
Comment (TheMarex) [inline include/engine/guidance/segment_list.hpp]: Why the copy here?

Comment (MoKob) [inline include/engine/guidance/segment_list.hpp]: Because typo performance bug. Thats why. Good catch!

File: include/engine/segment_information.hpp [hunk 5]
```diff
@@ -8,6 +8,8 @@
 #include "osrm/coordinate.hpp"
 #include <utility>
 
+#include <string>
```
Comment (TheMarex) [inline include/engine/segment_information.hpp]: Not needed by `SegmentInformation`.

File: include/extractor/edge_based_node.hpp [hunk 6]
```diff
@@ -80,6 +80,7 @@ struct EdgeBasedNode
     int forward_offset;          // prefix sum of the weight up the edge TODO: short must suffice
     int reverse_offset;          // prefix sum of the weight from the edge TODO: short must suffice
     unsigned packed_geometry_id; // if set, then the edge represents a packed geometry
+
```
Comment (TheMarex) [inline include/extractor/edge_based_node.hpp]: Remove whitespace change.

File: include/engine/routing_algorithms/routing_base.hpp [hunk 7]
```diff
@@ -614,6 +616,8 @@ template <class DataFacadeT, class Derived> class BasicRoutingInterface
         std::sort(reverse_entry_points.begin(), reverse_entry_points.end(), entry_point_comparator);
 
         NodeID last_id = SPECIAL_NODEID;
+        forward_core_heap.Clear();
+        reverse_core_heap.Clear();
```
Comment (daniel-j-h) [inline include/engine/routing_algorithms/routing_base.hpp]: Wait how does this come in here, or better why was this missing before? Bug?

**Comment by daniel-j-h** [comment]
Looks good code-wise. All the manual looping, erasing, shifting is a bit ugly (and the erasing somewhat dangerous), but that's because of the existing code / approach.
